File: templates/cli/lib/commands/generate.ts
import * as path from "path";
import { Command } from "commander";
import { ConfigType } from "./config.js";
import { localConfig } from "../config.js";
import { success, error, log, warn, actionRunner } from "../parser.js";
import {
  createGeneratorFromDetection,
  createGenerator,
  getSupportedLanguages,
  LanguageDetector,
  SupportedLanguage,
} from "./generators/index.js";
import { globalConfig } from "../config.js";
import {
  SDK_TITLE,
  SDK_TITLE_LOWER,
  EXECUTABLE_NAME,
  DEFAULT_ENDPOINT,
} from "../constants.js";
import {
  getAppwriteDependency,
  supportsServerSideMethods,
} from "../shared/typescript-type-utils.js";

type ServerSideOverride = "auto" | "true" | "false";

export interface GenerateCommandOptions {
  output: string;
  language?: string;
  serverSide?: ServerSideOverride;
}

const generateAction = async (
  options: GenerateCommandOptions,
): Promise<void> => {
  const project = localConfig.getProject();

  if (!project.projectId) {
    error(
      `No project found. Please run '${EXECUTABLE_NAME} init project' first.`,
    );
    process.exit(1);
  }

  // Determine the generator to use
  let generator;
  let detectedLanguage: string;

  const serverSideOverride: ServerSideOverride = options.serverSide ?? "auto";
  if (!["auto", "true", "false"].includes(serverSideOverride)) {
    error(`Invalid --server-side value: ${serverSideOverride}`);
    process.exit(1);
  }

  if (options.language) {
    // User explicitly specified a language
    if (!LanguageDetector.isSupported(options.language)) {
      const supported = getSupportedLanguages().join(", ");
      error(
        `Unsupported language: ${options.language}. Supported languages: ${supported}`,
      );
      process.exit(1);
    }
    generator = createGenerator(options.language as SupportedLanguage);
    detectedLanguage = options.language;
    log(`Using specified language: ${detectedLanguage}`);
  } else {
    // Auto-detect language
    try {
      const { generator: detectedGenerator, detection } =
        createGeneratorFromDetection();
      generator = detectedGenerator;
      detectedLanguage = detection.language;

      if (detection.confidence === "low") {
        warn(
          `Detected language '${detectedLanguage}' with low confidence (${detection.reason}). ` +
            `Use --language to specify explicitly.`,
        );
      } else {
        log(`Detected language: ${detectedLanguage} (${detection.reason})`);
      }
    } catch (err: any) {
      const supported = getSupportedLanguages().join(", ");
      error(
        `${err.message}\nUse --language to specify the target language. Supported: ${supported}`,
      );
      process.exit(1);
    }
  }

  if (typeof (generator as any).setServerSideOverride === "function") {
    (generator as any).setServerSideOverride(serverSideOverride);
  }

  const config: ConfigType = {
    projectId: project.projectId,
    projectName: project.projectName,
    endpoint:
      localConfig.getEndpoint() ||
      globalConfig.getEndpoint() ||
      DEFAULT_ENDPOINT,
    tablesDB: localConfig.getTablesDBs(),
    tables: localConfig.getTables(),
    databases: localConfig.getDatabases(),
    collections: localConfig.getCollections(),
  };

  const outputDir = options.output;
  const absoluteOutputDir = path.isAbsolute(outputDir)
    ? outputDir
    : path.join(process.cwd(), outputDir);

  log(
    `Generating type-safe ${detectedLanguage} SDK to ${absoluteOutputDir}...`,
  );

  try {
    const result = await generator.generate(config);
    await generator.writeFiles(absoluteOutputDir, result);

    const generatedFiles = generator.getGeneratedFilePaths(result);
    success(`Generated files:`);
    for (const file of generatedFiles) {
      console.log(`  - ${path.join(outputDir, file)}`);
    }

    // Show language-specific usage instructions
    if (detectedLanguage === "typescript") {
      const entities = config.tables?.length
        ? config.tables
        : config.collections;
      const firstEntity = entities?.[0];
      const dbId = firstEntity?.databaseId ?? "databaseId";
      const tableName = firstEntity?.name ?? "tableName";
      const formatAccessor = (value: string): string =>
        /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(value)
          ? `.${value}`
          : `[${JSON.stringify(value)}]`;
      const dbAccessor = formatAccessor(dbId);
      const tableAccessor = formatAccessor(tableName);

      console.log("");
      log(`Import the generated SDK in your project:`);
      console.log(
        `  import { databases } from "./${outputDir}/${SDK_TITLE_LOWER}/index.js";`,
      );
      console.log("");
      log(`Configure your client constants:`);
      console.log(
        `  set values in ./${outputDir}/${SDK_TITLE_LOWER}/constants.ts`,
      );
      console.log("");
      const appwriteDep = getAppwriteDependency();
      const supportsServerSide = supportsServerSideMethods(
        appwriteDep,
        serverSideOverride,
      );

      log(`Usage:`);
      if (supportsServerSide) {
        console.log(`  const mydb = databases.use(${JSON.stringify(dbId)});`);
        console.log(
          `  await mydb.use(${JSON.stringify(tableName)}).create({ ... });`,
        );
      } else {
        console.log(`  const mydb = databases${dbAccessor};`);
        console.log(`  await mydb${tableAccessor}.create({ ... });`);
      }
    }
  } catch (err: any) {
    error(`Failed to generate SDK: ${err.message}`);
    process.exit(1);
  }
};

export const generate = new Command("generate")
  .description(
    `Generate a type-safe SDK from your ${SDK_TITLE} project configuration`,
  )
  .option(
    "-o, --output <directory>",
    "Output directory for generated files (default: generated)",
    "generated",
  )
  .option(
    "-l, --language <language>",
    `Target language for SDK generation (supported: ${getSupportedLanguages().join(", ")})`,
  )
  .option(
    "--server-side <mode>",
    "Override server-side generation (auto|true|false)",
    "auto",
  )
  .action(actionRunner(generateAction));

File: templates/cli/internal/cmd/generate.go
//go:build !browser

package cmd

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/{{ sdk.gitUserName }}/{{ sdk.gitRepoName | caseDash }}/internal/app"
	"github.com/{{ sdk.gitUserName }}/{{ sdk.gitRepoName | caseDash }}/internal/config"
	"github.com/{{ sdk.gitUserName }}/{{ sdk.gitRepoName | caseDash }}/internal/generator"
	"github.com/{{ sdk.gitUserName }}/{{ sdk.gitRepoName | caseDash }}/internal/jsonx"
	"github.com/{{ sdk.gitUserName }}/{{ sdk.gitRepoName | caseDash }}/internal/output"
	"github.com/spf13/cobra"
)

// serverSideModes are the accepted --server values.
var serverSideModes = []string{"auto", "true", "false"}

func newGenerateCommand() *cobra.Command {
	var (
		outputDirectory string
		language        string
		serverSide      string
		importSource    string
		importExtension string
	)

	command := &cobra.Command{
		Use:   "generate",
		Short: "Generate a type-safe SDK from your Appwrite project configuration",
		RunE: func(command *cobra.Command, args []string) error {
			// An explicitly empty --import-extension is meaningful (it selects
			// CJS), so "was the flag given" is what decides whether to detect.
			var extension *string
			if command.Flags().Changed("import-extension") {
				extension = &importExtension
			}

			return runGenerate(command, outputDirectory, language, serverSide, importSource, extension)
		},
	}

	supported := make([]string, 0, len(generator.SupportedLanguages()))
	for _, name := range generator.SupportedLanguages() {
		supported = append(supported, string(name))
	}

	command.Flags().StringVarP(&outputDirectory, "output", "o", "generated",
		"Output directory for generated files")
	command.Flags().StringVarP(&language, "language", "l", "",
		"Target language for SDK generation (supported: "+strings.Join(supported, ", ")+")")
	command.Flags().StringVar(&serverSide, "server", "auto",
		"Override server-side generation (auto|true|false)")
	command.Flags().StringVar(&importSource, "appwrite-import-source", "",
		"Override Appwrite import source in generated files (e.g. node-appwrite, appwrite, @appwrite.io/console). "+
			"Auto-detected from package.json/deno.json if not provided.")
	command.Flags().StringVar(&importExtension, "import-extension", "",
		`Override import file extension in generated files (.js for ESM, empty string for CJS). `+
			`Auto-detected from package.json "type" field if not provided.`)

	return command
}

// resolveGenerator picks the generator, either from --language or by detection.
// resolveGenerateEndpoint fills in the endpoint the generated client will use.
func resolveGenerateEndpoint(configured string) string {
	if configured != "" {
		return configured
	}

	if global, err := preferences(); err == nil {
		if endpoint := global.CurrentValue(config.PreferenceEndpoint); endpoint != "" {
			return endpoint
		}
	}

	return config.DefaultEndpoint
}

func resolveGenerator(command *cobra.Command, requested, working string) (generator.Generator, string, error) {
	if requested != "" {
		created, err := generator.New(generator.Language(requested))
		if err != nil {
			return nil, "", err
		}
		output.Log(command.OutOrStdout(), "Using specified language: %s", requested)

		return created, requested, nil
	}

	created, detection, err := generator.NewFromDetection(working)
	if err != nil {
		supported := make([]string, 0, len(generator.SupportedLanguages()))
		for _, name := range generator.SupportedLanguages() {
			supported = append(supported, string(name))
		}

		return nil, "", fmt.Errorf(
			"%w\nUse --language to specify the target language. Supported: %s",
			err, strings.Join(supported, ", "))
	}

	// Low confidence means nothing but a matching file extension was found,
	// which is weak enough to be worth saying out loud.
	if detection.Confidence == generator.ConfidenceLow {
		output.Warn(command.OutOrStdout(),
			"Detected language '%s' with low confidence (%s). Use --language to specify explicitly.",
			detection.Language, detection.Reason)
	} else {
		output.Log(command.OutOrStdout(), "Detected language: %s (%s)",
			detection.Language, detection.Reason)
	}

	return created, string(detection.Language), nil
}

func runGenerate(
	command *cobra.Command,
	outputDirectory, language, serverSide, importSource string,
	importExtension *string,
) error {
	working, err := os.Getwd()
	if err != nil {
		return err
	}

	project, err := generator.LoadConfig(config.FindLocalPath())
	if err != nil || project.ProjectID == "" {
		return fmt.Errorf("no project found. Please run '%s init project' first", app.ExecutableName)
	}

	// The generated client hard-codes the endpoint it talks to, and `pull`
	// never writes an `endpoint` key -- so reading only the project config
	// emitted `export const ENDPOINT = ''` and produced a client that cannot
	// connect. Ports the localConfig || globalConfig || DEFAULT_ENDPOINT chain
	// in generate.ts:100.
	project.Endpoint = resolveGenerateEndpoint(project.Endpoint)

	if !contains(serverSideModes, serverSide) {
		return fmt.Errorf("invalid --server value: %s", serverSide)
	}

	created, name, err := resolveGenerator(command, language, working)
	if err != nil {
		return err
	}

	absolute := outputDirectory
	if !filepath.IsAbs(absolute) {
		absolute = filepath.Join(working, absolute)
	}

	output.Log(command.OutOrStdout(), "Generating type-safe %s SDK to %s...", name, absolute)

	result, err := created.Generate(project, generator.Options{
		ImportSource:    importSource,
		ImportExtension: importExtension,
		ServerSide:      serverSide,
	})
	if err != nil {
		return fmt.Errorf("failed to generate SDK: %w", err)
	}

	if err := generator.WriteFiles(absolute, generator.SDKDirectory, result); err != nil {
		return fmt.Errorf("failed to generate SDK: %w", err)
	}

	output.Success(command.OutOrStdout(), "Generated files:")
	for _, file := range generator.GeneratedFilePaths(generator.SDKDirectory) {
		command.Printf("  - %s\n", filepath.Join(outputDirectory, file))
	}

	if name == string(generator.LanguageTypeScript) {
		printTypeScriptUsage(command, project, outputDirectory, importExtension)
	}

	return nil
}

// printTypeScriptUsage prints the import and first-call hints.
func printTypeScriptUsage(
	command *cobra.Command,
	project generator.Config,
	outputDirectory string,
	importExtension *string,
) {
	extension := ""
	if importExtension != nil {
		extension = *importExtension
	} else {
		extension = generator.DetectImportExtension(".")
	}

	// Placeholders when the config declares nothing, so the hint still reads
	// as a complete example.
	databaseID, tableName := "databaseId", "tableName"
	if entities := project.Entities(); len(entities) > 0 {
		databaseID, tableName = entities[0].DatabaseID, entities[0].Name
	}

	out := command.OutOrStdout()

	command.Println()
	output.Log(out, "Import the generated SDK in your project:")
	command.Printf("  import { databases } from \"./%s/%s/index%s\";\n",
		outputDirectory, generator.SDKDirectory, extension)
	command.Println()
	output.Log(out, "Configure your SDK constants:")
	command.Printf("  set values in ./%s/%s/constants.ts\n", outputDirectory, generator.SDKDirectory)
	command.Println()
	output.Log(out, "Usage:")
	command.Printf("  const mydb = databases.use(%s);\n", quoteForHint(databaseID))
	command.Printf("  await mydb.use(%s).create({ ... });\n", quoteForHint(tableName))
}

// quoteForHint renders a JSON string literal, as JSON.stringify does in the
// TypeScript's usage hint.
func quoteForHint(value string) string {
	encoded, err := jsonx.Marshal(value)
	if err != nil {
		return `"` + value + `"`
	}

	return string(encoded)
}

func contains(values []string, value string) bool {
	for _, candidate := range values {
		if candidate == value {
			return true
		}
	}

	return false
}
